### PR TITLE
Liveness and readiness probes

### LINKED ISSUE #4005: Support division of readiness and liveness probes in /health endpoint
Kubernetes supports both readiness and liveness probes, currently Micronaut's management module only really supports liveness.

We should add qualifiers to define health indicators as either `@Readiness` or `@Liveness` and define a selector on the health endpoint to expose these individually `/health` can continue to map to liveness.

Micronaut launch's `k8s.yml` should then be updated to point to the individual probes.

```yaml
         readinessProbe:
            httpGet:
              path: /health/readiness
              port: 8080
            initialDelaySeconds: 5
            timeoutSeconds: 3
          livenessProbe:
            httpGet:
              path: /health/liveness
              port: 8080
            initialDelaySeconds: 5
            timeoutSeconds: 3
            failureThreshold: 10
```

### PR DESCRIPTION
**Pull request by pgressa**
Fixes #4005

### CODE REVIEW

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java [hunk 1]
```diff
@@ -79,8 +90,39 @@ public HealthEndpoint(HealthAggregator<HealthResult> healthAggregator,
...
     public Single<HealthResult> getHealth(@Nullable Principal principal) {
         HealthLevelOfDetail detail = levelOfDetail(principal);
+        // due to backward comaptibility all HealthIndicators that are not qualified with Readiness
+        // e.g. are Liveness or without qualifier are returned by this resource
         return Single.fromPublisher(
-                healthAggregator.aggregate(healthIndicators, detail)
+                healthAggregator.aggregate(
+                        Arrays.stream(healthIndicators).
+                                filter(healthIndicator ->
+                                        !Arrays.asList(readinessHealthIndicators).contains(healthIndicator))
```
Comment (pgressa) [inline management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java]: I don't like this line but I expect that Array.asList will be somehow optimised by JIT

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java [hunk 2]
```diff
@@ -79,8 +90,39 @@ public HealthEndpoint(HealthAggregator<HealthResult> healthAggregator,
...
+        );
+    }
+
+    /**
+     * @param principal Authenticated user
+     * @param selector HealthEndpointSelector
+     * @return The health information as a {@link Single}
+     */
+    @Read
+    public Single<HealthResult> getHealth(@Nullable Principal principal, @Selector HealthEndpointSelector selector) {
```
Comment (pgressa) [inline management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java]: Not sure whether using enum in selector isn't too much defensive

File: management/src/main/java/io/micronaut/management/health/indicator/Liveness.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * @author Pavol Gressa
+ * @since 2.1
+ */
+@Qualifier
+@Documented
+@Retention(RUNTIME)
+public @interface Liveness{
```
Comment (pgressa) [inline management/src/main/java/io/micronaut/management/health/indicator/Liveness.java]: `Liveness` or `LivenessHealthIndicator`?

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java (hunk 2 above)
Comment (graemerocher) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java]: I think an enum is fine.

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Options for {@link io.micronaut.management.endpoint.health.HealthEndpoint} selector that aggregates
+ * {@link io.micronaut.management.health.indicator.HealthIndicator} according to {@link io.micronaut.management.health.indicator.Liveness}
+ * respectively {@link io.micronaut.management.health.indicator.Readiness} qualifiers.
+ *
+ * @author Pavol Gressa
+ * @since 2.1
+ */
+public enum HealthEndpointSelector {
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java]: I would move this to the package `io.micronaut.management.health.indicator` and rename to `HealthCheckType`

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package io.micronaut.management.endpoint.health;
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java]: Add license (can be done with `./gradlew spotlessApply`

File: management/src/main/java/io/micronaut/management/health/indicator/Liveness.java (hunk 3 above)
Comment (graemerocher) [reply management/src/main/java/io/micronaut/management/health/indicator/Liveness.java]: Liveness works

File: management/src/main/java/io/micronaut/management/health/indicator/Liveness.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package io.micronaut.management.health.indicator;
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Liveness.java]: add license

File: management/src/main/java/io/micronaut/management/health/indicator/Liveness.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+import javax.inject.Qualifier;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * @author Pavol Gressa
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Liveness.java]: add javadoc description

File: management/src/main/java/io/micronaut/management/health/indicator/Readiness.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package io.micronaut.management.health.indicator;
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Readiness.java]: add license

File: management/src/main/java/io/micronaut/management/health/indicator/Readiness.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+import javax.inject.Qualifier;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * @author Pavol Gressa
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Readiness.java]: Add javadoc description

File: management/src/main/java/io/micronaut/management/health/indicator/Readiness.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ * @since 2.1
+ */
+@Qualifier
+@Documented
+@Retention(RUNTIME)
+public @interface Readiness {
```
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Readiness.java]: `@Target` should be TYPE, METHOD and ANNOTATION_TYPE

File: management/src/main/java/io/micronaut/management/health/indicator/Readiness.java (hunk 8 above)
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Readiness.java]: put these in a subpackage called `io.micronaut.management.health.indicator.annotation`

File: management/src/main/java/io/micronaut/management/health/indicator/Liveness.java (hunk 3 above)
Comment (graemerocher) [inline management/src/main/java/io/micronaut/management/health/indicator/Liveness.java]: `@Target` should be TYPE, METHOD and ANNOTATION_TYPE

**Review by graemerocher — CHANGES_REQUESTED** [review]
Nice work, needs checkstyle and javadoc cleanup, but Good start.

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java (hunk 1 above)
Comment (graemerocher) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java]: You should probably at least assign the list to a local variable outside the scope. I would also move the filter to the constructor of the class as there is no need to do this calculation for each request

Comment (pgressa) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java]: yeah, that's right

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java (hunk 5 above)
Comment (pgressa) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java]: thanks for hint, will remember that, completely forgot for licences

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java (hunk 4 above)
Comment (pgressa) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpointSelector.java]: OK

File: management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java [hunk 11]
```diff
@@ -72,15 +71,22 @@
...
     public HealthEndpoint(HealthAggregator<HealthResult> healthAggregator,
                           HealthIndicator[] healthIndicators,
-                          @Liveness HealthIndicator[] livenessHealthIndicators,
-                          @Readiness HealthIndicator[] readinessHealthIndicators) {
+                          @Liveness HealthIndicator[] livenessHealthIndicators) {
         this.healthAggregator = healthAggregator;
-        this.healthIndicators = healthIndicators;
         this.livenessHealthIndicators = livenessHealthIndicators;
-        this.readinessHealthIndicators = readinessHealthIndicators;
+        this.readinessHealthIndicators = getReadinessHealthIndicators(healthIndicators, livenessHealthIndicators);
```
Comment (pgressa) [inline management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java:83]: so this basically means that `/health` and `/health/readiness` returns the same payload

Comment (pgressa) [reply management/src/main/java/io/micronaut/management/endpoint/health/HealthEndpoint.java:83]: no it does not :) .. health returns all healthindicators, and health/readiness returns all @Readiness and all not qualified health indicators

File: src/main/docs/guide/management/providedEndpoints/healthEndpoint.adoc [hunk 12]
```diff
@@ -69,7 +69,7 @@ The above will return api:http.HttpStatus#OK[] (200) even when the api:health.He
 
 == Customization
 
-The health endpoint consists of a health aggregator and any number of health indicators. To add a health indicator, simply create a bean class that implements link:{api}/io/micronaut/management/health/indicator/HealthIndicator.html[HealthIndicator]. A base class link:{api}/io/micronaut/management/health/indicator/AbstractHealthIndicator.html[AbstractHealthIndicator] has been provided for you to extend to make the process easier.
+The health endpoint consists of a health aggregator and any number of health indicators. To add a health indicator, simply create a bean class that implements link:{api}/io/micronaut/management/health/indicator/HealthIndicator.html[HealthIndicator]. It is recommended to also use either link:{api}/io/micronaut/health/indicator/annotation/Liveness.html[@Liveness] or link:{api}/io/micronaut/health/indicator/annotation/Readiness.html[@Readiness] qualifier. If no qualifier is used then the health incidator is part of /health and /health/readiness endpoints. A base class link:{api}/io/micronaut/management/health/indicator/AbstractHealthIndicator.html[AbstractHealthIndicator] has been provided for you to extend to make the process easier.
```
Comment (pgressa) [inline src/main/docs/guide/management/providedEndpoints/healthEndpoint.adoc:72]: This will need better phrasing.

**Comment by alvarosanchez** [comment]
Are we planning to add a default `@Liveness` `HealthIndicator`?

**Comment by alvarosanchez** [comment]
> Are we planning to add a default `@Liveness` `HealthIndicator`?

To elaborate more on this: I am suggesting we create a `ServiceReadyHealthIndicator` that returns `OK` only after the `ServiceReadyEvent` has been fired.

**Comment by graemerocher** [comment]
Yes, in addition this event listener should be the last in the chain (sort using `Ordered`)

**Comment by graemerocher** [comment]
@pgressa please add the health indicator as indicated by @alvarosanchez 

Need to get this merged soon.

**Comment by pgressa** [comment]
@graemerocher @alvarosanchez what about https://github.com/micronaut-projects/micronaut-core/pull/4012/commits/486a3499fd69c0d930628e73340e718892021669#diff-ce502d8fb2599176efd4616c2a159e79 ? I had to update also test, namely that app context has set the application name, what makes it imho fragile as that property has to be configured otherwise the ServiceReadyEvent is not emitted https://github.com/micronaut-projects/micronaut-core/blob/2.0.x/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java#L496 . I can be of course mistaken as by default a name is always generated.

**Comment by graemerocher** [comment]
looks good to me, @alvarosanchez ?

**Review by alvarosanchez — APPROVED** [review]
Looks great! This is worth to be mentioned in the what's new section.

**Comment by graemerocher** [comment]
@pgressa please add an entry to `whatsNew.adoc`... thanks

**Comment by pgressa** [comment]
@graemerocher @alvarosanchez please check https://github.com/micronaut-projects/micronaut-core/pull/4012/commits/d195d6662beca30ac9857bf66557a12d75e43bdd .. will squash afterwards

**Pull request merged**